fuzz/fuzz_targets/value_binop.rs

```diff
@@ -15,13 +19,111 @@ enum BinOp {
+            }
+            count.set(count.get() + 1);
+            Ok(())
+        })?;
+    }
+
+    let result = Ok(
+        match (u64::from(<u32 as arbitrary::Arbitrary>::arbitrary(u)?) * 11u64) >> 32 {
+            0u64 => {
+                let length = <u8 as arbitrary::Arbitrary>::arbitrary(u)?;
+                let mut list = Vec::with_capacity(length as usize);
+                for _ in 0..length {
+                    list.push(arbitrary_value(u)?);
+                }
+                Value::List(Arc::new(list))
+            }
+            1u64 => {
+                let length = <u8 as arbitrary::Arbitrary>::arbitrary(u)?;
+                let mut map = HashMap::with_capacity(length as usize);
+                for _ in 0..length {
+                    map.insert(arbitrary_key(u)?, arbitrary_value(u)?);
+                }
+                Value::Map(Map { map: Arc::new(map) })
+            }
+            2u64 => Value::Int(arbitrary::Arbitrary::arbitrary(u)?),
+            3u64 => Value::UInt(arbitrary::Arbitrary::arbitrary(u)?),
+            4u64 => Value::Float(arbitrary::Arbitrary::arbitrary(u)?),
+            5u64 => Value::String(arbitrary::Arbitrary::arbitrary(u)?),
+            6u64 => Value::Bytes(arbitrary::Arbitrary::arbitrary(u)?),
+            7u64 => Value::Bool(arbitrary::Arbitrary::arbitrary(u)?),
+            8u64 => Value::Duration(chrono::Duration::nanoseconds(
+                arbitrary::Arbitrary::arbitrary(u)?,
+            )),
+            9u64 => Value::Timestamp(
+                chrono::Utc
+                    .timestamp_nanos(arbitrary::Arbitrary::arbitrary(u)?)
+                    .into(),
+            ),
+            10u64 => Value::Null,
+            _ => unreachable!(),
+        },
+    );
+
+    if guard_against_recursion {
+        RECURSIVE_COUNT_Value.with(|count| {
+            count.set(count.get() - 1);
+        });
+    }
+    result
+}
+
+fn arbitrary_key(u: &mut arbitrary::Unstructured<'_>) -> arbitrary::Result<Key> {
+    let guard_against_recursion = u.is_empty();
+    if guard_against_recursion {
+        RECURSIVE_COUNT_Value.with(|count| {
+            if count.get() > 0 {
+                return Err(arbitrary::Error::NotEnoughData);
+            }
+            count.set(count.get() + 1);
+            Ok(())
+        })?;
+    }
+
+    let result = Ok(
+        match (u64::from(<u32 as arbitrary::Arbitrary>::arbitrary(u)?) * 4u64) >> 32 {
+            0u64 => Key::Int(arbitrary::Arbitrary::arbitrary(u)?),
+            1u64 => Key::Uint(arbitrary::Arbitrary::arbitrary(u)?),
+            2u64 => Key::Bool(arbitrary::Arbitrary::arbitrary(u)?),
+            3u64 => Key::String(Arc::new(arbitrary::Arbitrary::arbitrary(u)?)),
+            _ => unreachable!(),
+        },
+    );
+
+    if guard_against_recursion {
+        RECURSIVE_COUNT_Value.with(|count| {
+            count.set(count.get() - 1);
+        });
+    }
+    result
+}
```

Comment:
`arbitrary_key` is not recursively called AFAICT, so what is the idea behind the recursive guards?